File: src/cli/output.py
"""Console output helpers with scoped indentation."""

from contextlib import contextmanager
from typing import Iterator

from rich.console import Console


class Output:
    """Console output helpers with scoped indentation.

    Usage:
        out = Output()
        out.section("Starting process...")
        with out.indent():
            out.info("Step 1")
            with out.indent():
                out.success("Done")
    """

    def __init__(self, console: Console | None = None) -> None:
        self.console = console or Console()
        self._indent_level = 0

    @property
    def _prefix(self) -> str:
        return " " * self._indent_level

    @contextmanager
    def indent(self, spaces: int = 2) -> Iterator[None]:
        """Context manager for scoped indentation."""
        self._indent_level += spaces
        try:
            yield
        finally:
            self._indent_level -= spaces

    @contextmanager
    def operation(self, title: str, color: str = "blue") -> Iterator[None]:
        """Print a section header and indent the block."""
        self.section(title, color)
        with self.indent():
            yield

    def error(self, msg: str) -> None:
        """Print an error message in red."""
        self.console.print(f"{self._prefix}[red]Error:[/red] {msg}")

    def warning(self, msg: str) -> None:
        """Print a warning message in yellow."""
        self.console.print(f"{self._prefix}[yellow]Warning:[/yellow] {msg}")

    def hint(self, msg: str) -> None:
        """Print a hint message in yellow."""
        self.console.print(f"{self._prefix}[yellow]Hint:[/yellow] {msg}")

    def success(self, msg: str) -> None:
        """Print a success message with green checkmark."""
        self.console.print(f"{self._prefix}[green]✓[/green] {msg}")

    def failure(self, msg: str) -> None:
        """Print a failure message with red X."""
        self.console.print(f"{self._prefix}[red]✗[/red] {msg}")

    def section(self, title: str, color: str = "blue") -> None:
        """Print a bold section header."""
        self.console.print(f"{self._prefix}[bold {color}]{title}[/bold {color}]")

    def dim(self, msg: str) -> None:
        """Print a dimmed message."""
        self.console.print(f"{self._prefix}[dim]{msg}[/dim]")

    def info(self, msg: str) -> None:
        """Print an info message (no special formatting)."""
        self.console.print(f"{self._prefix}{msg}")

    def styled(self, msg: str) -> None:
        """Print a message with Rich markup (caller handles formatting)."""
        self.console.print(f"{self._prefix}{msg}")


# Module-level instance for convenience
out = Output()

File: src/cli/output.cpp
/*
    SPDX-FileCopyrightText: 2024-2026 KDE Community
    SPDX-License-Identifier: GPL-3.0-or-later
*/

#include "output.h"
#include "rang.hpp"

#include <algorithm>
#include <iomanip>

namespace Kapsule {

Output &out()
{
    static Output instance;
    return instance;
}

void Output::printPrefix(int extraIndent)
{
    int total = m_indentLevel + extraIndent;
    for (int i = 0; i < total; ++i) {
        m_stream << ' ';
    }
}

void Output::error(std::string_view msg)
{
    printPrefix();
    m_stream << rang::fg::red << "Error:" << rang::fg::reset << " " << msg << '\n';
}

void Output::warning(std::string_view msg)
{
    printPrefix();
    m_stream << rang::fg::yellow << "Warning:" << rang::fg::reset << " " << msg << '\n';
}

void Output::hint(std::string_view msg)
{
    printPrefix();
    m_stream << rang::fg::yellow << "Hint:" << rang::fg::reset << " " << msg << '\n';
}

void Output::success(std::string_view msg)
{
    printPrefix();
    m_stream << rang::fg::green << "✓" << rang::fg::reset << " " << msg << '\n';
}

void Output::failure(std::string_view msg)
{
    printPrefix();
    m_stream << rang::fg::red << "✗" << rang::fg::reset << " " << msg << '\n';
}

void Output::section(std::string_view title)
{
    printPrefix();
    m_stream << rang::style::bold << rang::fg::blue << title
             << rang::fg::reset << rang::style::reset << '\n';
}

void Output::dim(std::string_view msg)
{
    printPrefix();
    m_stream << rang::style::dim << msg << rang::style::reset << '\n';
}

void Output::info(std::string_view msg)
{
    printPrefix();
    m_stream << msg << '\n';
}

void Output::print(MessageType type, std::string_view msg, int extraIndent)
{
    // Calculate total indent (current level + extra from message)
    int savedIndent = m_indentLevel;
    m_indentLevel += extraIndent * 2;  // 2 spaces per indent level

    switch (type) {
    case MessageType::Info:
        info(msg);
        break;
    case MessageType::Success:
        success(msg);
        break;
    case MessageType::Warning:
        warning(msg);
        break;
    case MessageType::Error:
        error(msg);
        break;
    case MessageType::Dim:
        dim(msg);
        break;
    case MessageType::Hint:
        hint(msg);
        break;
    }

    m_indentLevel = savedIndent;
}

void Output::progress(std::string_view description, int current, int total)
{
    printPrefix();

    if (total > 0) {
        // Determinate progress
        int percent = (current * 100) / total;
        int barWidth = 30;
        int filled = (current * barWidth) / total;

        m_stream << rang::fg::cyan << description << rang::fg::reset << " [";
        for (int i = 0; i < barWidth; ++i) {
            if (i < filled) {
                m_stream << rang::fg::green << "█" << rang::fg::reset;
            } else {
                m_stream << rang::style::dim << "░" << rang::style::reset;
            }
        }
        m_stream << "] " << percent << "%\r";
    } else {
        // Indeterminate progress (spinner-like)
        static const char *spinChars = "⠋⠙⠹⠸⠼⠴⠦⠧⠇⠏";
        // Each spinner char is 3 bytes in UTF-8
        int idx = (current % 10) * 3;
        m_stream << rang::fg::cyan << std::string_view(spinChars + idx, 3)
                 << " " << description << rang::fg::reset << "\r";
    }
    m_stream.flush();
}

void Output::indent(int spaces)
{
    m_indentLevel += spaces;
}

void Output::dedent(int spaces)
{
    m_indentLevel = std::max(0, m_indentLevel - spaces);
}

} // namespace Kapsule
